File: scripts/github-actions/generate_daily_featured_stocks.py
#!/usr/bin/env python3
"""
DailyFeaturedStock自動生成スクリプト

株価変動率、取引高、時価総額から機械的に銘柄を3カテゴリに分類：
- surge（短期急騰）: 週間上昇率+5%以上（時価総額はスコアで加点）
- stable（中長期安定）: ボラティリティ15%以下（時価総額はスコアで加点）
- trending（話題）: 出来高比率1.5倍以上（時価総額はスコアで加点）

毎日朝に実行され、各カテゴリTop 3を選出（合計9銘柄）
"""

import os
import sys
import uuid
from datetime import datetime, timezone

import psycopg2
import psycopg2.extras


def get_database_url() -> str:
    """データベースURLを取得"""
    url = os.environ.get("DATABASE_URL")
    if not url:
        print("Error: DATABASE_URL environment variable not set")
        sys.exit(1)
    return url


def fetch_stocks_with_prices(conn) -> list[dict]:
    """DBから株価データを持つ銘柄一覧を取得"""
    with conn.cursor() as cur:
        cur.execute('''
            SELECT
                id,
                "tickerCode",
                name,
                "marketCap",
                "weekChangeRate",
                "volatility",
                "volumeRatio"
            FROM "Stock"
            WHERE "priceUpdatedAt" IS NOT NULL
              AND "latestPrice" IS NOT NULL
              AND "marketCap" IS NOT NULL
            ORDER BY "marketCap" DESC NULLS LAST
        ''')
        rows = cur.fetchall()

    stocks = []
    for row in rows:
        stocks.append({
            "id": row[0],
            "tickerCode": row[1],
            "name": row[2],
            "marketCap": float(row[3]) if row[3] else 0,
            "weekChangeRate": float(row[4]) if row[4] else 0,
            "volatility": float(row[5]) if row[5] else None,
            "volumeRatio": float(row[6]) if row[6] else None,
        })
    return stocks


def market_cap_score(market_cap: float) -> float:
    """時価総額を0-100のスコアに変換（10000億円以上で100）"""
    return min(100, market_cap / 100)


def market_cap_label(market_cap: float) -> str:
    """時価総額を表示用ラベルに変換"""
    if market_cap >= 10000:
        return f"{market_cap / 10000:.1f}兆円"
    return f"{market_cap:.0f}億円"


def calculate_surge_stocks(stocks: list[dict]) -> list[dict]:
    """surge（短期急騰）銘柄を抽出"""
    candidates = []

    for stock in stocks:
        change_rate = stock["weekChangeRate"]
        if change_rate >= 5.0:
            # 上昇率スコア(70%) + 時価総額スコア(30%)
            primary = min(100, change_rate * 5)
            cap = market_cap_score(stock["marketCap"])
            composite = primary * 0.7 + cap * 0.3
            candidates.append({
                "stock": stock,
                "changeRate": change_rate,
                "compositeScore": composite,
            })

    candidates.sort(key=lambda x: x["compositeScore"], reverse=True)

    results = []
    for idx, candidate in enumerate(candidates[:3]):
        stock = candidate["stock"]
        cap_label = market_cap_label(stock["marketCap"])

        reason = f"この1週間で株価が{candidate['changeRate']:.1f}%上昇しています（時価総額{cap_label}）"

        results.append({
            "stockId": stock["id"],
            "category": "surge",
            "categoryPosition": idx + 1,
            "reason": reason,
            "score": min(100, int(candidate["compositeScore"])),
        })

    print(f"Surge: {len(results)} stocks selected")
    return results


def calculate_stable_stocks(stocks: list[dict]) -> list[dict]:
    """stable（中長期安定）銘柄を抽出"""
    candidates = []

    for stock in stocks:
        volatility = stock["volatility"]
        if volatility is None:
            continue

        if volatility <= 15.0:
            # 安定性スコア(60%) + 時価総額スコア(40%)
            primary = max(0, (20 - volatility) / 20 * 100)
            cap = market_cap_score(stock["marketCap"])
            composite = primary * 0.6 + cap * 0.4
            candidates.append({
                "stock": stock,
                "volatility": volatility,
                "compositeScore": composite,
            })

    candidates.sort(key=lambda x: x["compositeScore"], reverse=True)

    results = []
    for idx, candidate in enumerate(candidates[:3]):
        stock = candidate["stock"]
        cap_label = market_cap_label(stock["marketCap"])

        reason = f"安定した値動きで、初心者に最適な銘柄です（時価総額{cap_label}、変動率{candidate['volatility']:.1f}%）"

        results.append({
            "stockId": stock["id"],
            "category": "stable",
            "categoryPosition": idx + 1,
            "reason": reason,
            "score": min(100, int(candidate["compositeScore"])),
        })

    print(f"Stable: {len(results)} stocks selected")
    return results


def calculate_trending_stocks(stocks: list[dict]) -> list[dict]:
    """trending（話題）銘柄を抽出"""
    candidates = []

    for stock in stocks:
        volume_ratio = stock["volumeRatio"]
        if volume_ratio is None:
            continue

        if volume_ratio >= 1.5:
            # 出来高比率スコア(70%) + 時価総額スコア(30%)
            primary = min(100, volume_ratio * 25)
            cap = market_cap_score(stock["marketCap"])
            composite = primary * 0.7 + cap * 0.3
            candidates.append({
                "stock": stock,
                "volumeRatio": volume_ratio,
                "compositeScore": composite,
            })

    candidates.sort(key=lambda x: x["compositeScore"], reverse=True)

    results = []
    for idx, candidate in enumerate(candidates[:3]):
        stock = candidate["stock"]
        cap_label = market_cap_label(stock["marketCap"])

        reason = f"直近3日で取引が活発になっている注目銘柄です（取引高{candidate['volumeRatio']:.1f}倍、時価総額{cap_label}）"

        results.append({
            "stockId": stock["id"],
            "category": "trending",
            "categoryPosition": idx + 1,
            "reason": reason,
            "score": min(100, int(candidate["compositeScore"])),
        })

    print(f"Trending: {len(results)} stocks selected")
    return results


def save_daily_featured_stocks(conn, featured_stocks: list[dict]) -> None:
    """DailyFeaturedStockテーブルに保存"""
    if not featured_stocks:
        print("No stocks to save")
        return

    today = datetime.now(timezone.utc).replace(hour=0, minute=0, second=0, microsecond=0)

    with conn.cursor() as cur:
        cur.execute('''
            DELETE FROM "DailyFeaturedStock"
            WHERE date = %s
        ''', (today,))
        deleted_count = cur.rowcount
        print(f"Deleted {deleted_count} existing records for {today.date()}")

        for idx, fs in enumerate(featured_stocks):
            cur.execute('''
                INSERT INTO "DailyFeaturedStock" (id, date, "stockId", category, position, reason, score, "createdAt")
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
            ''', (
                str(uuid.uuid4().hex[:25]),
                today,
                fs["stockId"],
                fs["category"],
                idx + 1,
                fs["reason"],
                fs["score"],
                datetime.now(timezone.utc),
            ))

    conn.commit()
    print(f"Saved {len(featured_stocks)} featured stocks for {today.date()}")


def main():
    print("=" * 60)
    print("DailyFeaturedStock Generation (Python)")
    print("=" * 60)
    print(f"Time: {datetime.now().isoformat()}")
    print()

    db_url = get_database_url()
    conn = psycopg2.connect(db_url)

    try:
        print("Fetching stocks with price data...")
        stocks = fetch_stocks_with_prices(conn)
        print(f"Found {len(stocks)} stocks with price data")

        if not stocks:
            print("No stocks with sufficient data. Exiting.")
            return

        print()
        print("Calculating featured stocks...")

        surge_stocks = calculate_surge_stocks(stocks)
        stable_stocks = calculate_stable_stocks(stocks)
        trending_stocks = calculate_trending_stocks(stocks)

        all_featured = surge_stocks + stable_stocks + trending_stocks

        if not all_featured:
            print("No stocks matched criteria today")
            return

        print()
        print("Saving to database...")
        save_daily_featured_stocks(conn, all_featured)

        print()
        print("=" * 60)
        print("DailyFeaturedStock generation completed")
        print("=" * 60)
        print(f"Total featured stocks: {len(all_featured)}")
        print(f"  - Surge: {len(surge_stocks)}")
        print(f"  - Stable: {len(stable_stocks)}")
        print(f"  - Trending: {len(trending_stocks)}")
        print("=" * 60)

    finally:
        conn.close()


if __name__ == "__main__":
    main()

File: scripts/github-actions/generate-daily-featured-stocks.ts
#!/usr/bin/env npx tsx
/**
 * DailyFeaturedStock自動生成スクリプト
 *
 * 株価変動率、取引高、時価総額から機械的に銘柄を3カテゴリに分類：
 * - surge（短期急騰）: 7日間上昇率+5%以上、時価総額1000億円以上
 * - stable（中長期安定）: 時価総額5000億円以上 & ボラティリティ15%以下
 * - trending（話題）: 直近3日の取引高が4〜30日前平均の1.5倍以上
 *
 * 毎日朝7時（JST）に実行され、各カテゴリTop 3を選出（合計9銘柄）
 */

import { PrismaClient } from "@prisma/client"
import dayjs from "dayjs"
import utc from "dayjs/plugin/utc"
import { fetchHistoricalPrices } from "../../lib/stock-price-fetcher"

dayjs.extend(utc)

const prisma = new PrismaClient()

interface PriceData {
  date: string
  close: number
  volume: number
}

interface StockWithPrices {
  id: string
  tickerCode: string
  name: string
  marketCap: number // 時価総額（億円）
  prices: PriceData[]
}

interface FeaturedStock {
  stockId: string
  category: "surge" | "stable" | "trending"
  categoryPosition: number
  reason: string
  score: number
}

async function getStocksWithPrices(): Promise<StockWithPrices[]> {
  // 銘柄マスタを取得（時価総額が設定されている銘柄のみ）
  const stocks = await prisma.stock.findMany({
    where: { marketCap: { not: null } },
    orderBy: { marketCap: "desc" },
    select: {
      id: true,
      tickerCode: true,
      name: true,
      marketCap: true,
    },
  })
  console.log(`Found ${stocks.length} stocks in master`)

  if (stocks.length === 0) return []

  // Stooq APIで株価を取得
  console.log(`Fetching prices for ${stocks.length} stocks from Stooq API...`)

  const stocksWithPrices: StockWithPrices[] = []

  // バッチ処理
  const batchSize = 50
  for (let i = 0; i < stocks.length; i += batchSize) {
    const batchStocks = stocks.slice(i, i + batchSize)

    const results = await Promise.allSettled(
      batchStocks.map(async (stock) => {
        try {
          // 1ヶ月分のデータを取得
          const historicalData = await fetchHistoricalPrices(stock.tickerCode, "1m")

          if (!historicalData || historicalData.length < 7) return null

          // 新しい順にソート
          const prices: PriceData[] = historicalData
            .filter((d) => d.close && d.volume)
            .map((d) => ({
              date: d.date,
              close: d.close,
              volume: d.volume,
            }))
            .sort((a, b) => b.date.localeCompare(a.date))

          if (prices.length < 7) return null

          return {
            id: stock.id,
            tickerCode: stock.tickerCode,
            name: stock.name,
            marketCap: Number(stock.marketCap),
            prices,
          }
        } catch {
          return null
        }
      })
    )

    for (const result of results) {
      if (result.status === "fulfilled" && result.value) {
        stocksWithPrices.push(result.value)
      }
    }
  }

  console.log(`Found ${stocksWithPrices.length} stocks with sufficient price data`)
  return stocksWithPrices
}

function calculateSurgeStocks(stocks: StockWithPrices[]): FeaturedStock[] {
  /**
   * surge（短期急騰）銘柄を抽出
   *
   * 条件:
   * - 7日間の株価上昇率: +5%以上
   * - 時価総額: 1000億円以上
   */
  const surgeCandidates: { stock: StockWithPrices; changeRate: number; latestPrice: number }[] = []

  for (const stock of stocks) {
    // 時価総額1000億円以上の銘柄のみ
    if (stock.marketCap < 1000) continue

    const prices = stock.prices
    if (prices.length < 7) continue

    // 7日間の上昇率を計算
    const latestPrice = prices[0].close
    const weekAgoPrice = prices[6].close

    if (weekAgoPrice === 0) continue

    const changeRate = ((latestPrice - weekAgoPrice) / weekAgoPrice) * 100

    if (changeRate >= 5.0) {
      surgeCandidates.push({
        stock,
        changeRate,
        latestPrice,
      })
    }
  }

  // 上昇率が高い順にソート
  surgeCandidates.sort((a, b) => b.changeRate - a.changeRate)

  // Top 3を選出
  const topSurge = surgeCandidates.slice(0, 3)

  const results: FeaturedStock[] = topSurge.map((candidate, idx) => {
    const marketCapLabel = candidate.stock.marketCap >= 10000
      ? `${(candidate.stock.marketCap / 10000).toFixed(1)}兆円`
      : `${candidate.stock.marketCap.toFixed(0)}億円`
    const reason = `この1週間で株価が${candidate.changeRate.toFixed(1)}%上昇しています。時価総額${marketCapLabel}の安定した企業です`

    return {
      stockId: candidate.stock.id,
      category: "surge",
      categoryPosition: idx + 1,
      reason,
      score: Math.min(100, Math.floor(candidate.stock.marketCap / 100)), // 時価総額ベースのスコア
    }
  })

  console.log(`Surge: ${results.length} stocks selected`)
  return results
}

function calculateStableStocks(stocks: StockWithPrices[]): FeaturedStock[] {
  /**
   * stable（中長期安定）銘柄を抽出
   *
   * 条件:
   * - 時価総額: 5000億円以上
   * - 30日間のボラティリティ: 15%以下
   */
  const stableCandidates: { stock: StockWithPrices; volatility: number }[] = []

  for (const stock of stocks) {
    // 時価総額5000億円以上の大企業のみ
    if (stock.marketCap < 5000) continue

    const prices = stock.prices
    if (prices.length < 30) continue

    // ボラティリティを計算（標準偏差 / 平均）
    const closePrices = prices.map((p) => p.close)
    const avgPrice = closePrices.reduce((a, b) => a + b, 0) / closePrices.length
    const squareDiffs = closePrices.map((p) => Math.pow(p - avgPrice, 2))
    const stdDev = Math.sqrt(squareDiffs.reduce((a, b) => a + b, 0) / closePrices.length)

    if (avgPrice === 0) continue

    const volatility = (stdDev / avgPrice) * 100

    if (volatility <= 15.0) {
      stableCandidates.push({
        stock,
        volatility,
      })
    }
  }

  // 時価総額が大きい順にソート
  stableCandidates.sort((a, b) => b.stock.marketCap - a.stock.marketCap)

  // Top 3を選出
  const topStable = stableCandidates.slice(0, 3)

  const results: FeaturedStock[] = topStable.map((candidate, idx) => {
    const marketCapLabel = candidate.stock.marketCap >= 10000
      ? `${(candidate.stock.marketCap / 10000).toFixed(1)}兆円`
      : `${candidate.stock.marketCap.toFixed(0)}億円`
    const reason = `安定した値動きで、初心者に最適な銘柄です（時価総額${marketCapLabel}、変動率${candidate.volatility.toFixed(1)}%）`

    return {
      stockId: candidate.stock.id,
      category: "stable",
      categoryPosition: idx + 1,
      reason,
      score: Math.min(100, Math.floor(candidate.stock.marketCap / 100)), // 時価総額ベースのスコア
    }
  })

  console.log(`Stable: ${results.length} stocks selected`)
  return results
}

function calculateTrendingStocks(stocks: StockWithPrices[]): FeaturedStock[] {
  /**
   * trending（話題）銘柄を抽出
   *
   * 条件:
   * - 直近3日の平均取引高 > 4〜30日前の平均取引高 × 1.5倍
   * - 時価総額: 500億円以上
   */
  const trendingCandidates: { stock: StockWithPrices; volumeRatio: number }[] = []

  for (const stock of stocks) {
    // 時価総額500億円以上の銘柄のみ
    if (stock.marketCap < 500) continue

    const prices = stock.prices
    if (prices.length < 30) continue

    // 直近3日の平均取引高
    const recentVolumes = prices.slice(0, 3).map((p) => p.volume).filter((v) => v > 0)
    if (recentVolumes.length === 0) continue
    const recentAvgVolume = recentVolumes.reduce((a, b) => a + b, 0) / recentVolumes.length

    // 4〜30日前の平均取引高（重複を避ける）
    const olderVolumes = prices.slice(3, 30).map((p) => p.volume).filter((v) => v > 0)
    if (olderVolumes.length === 0) continue
    const olderAvgVolume = olderVolumes.reduce((a, b) => a + b, 0) / olderVolumes.length

    if (olderAvgVolume === 0) continue

    // 取引高増加率
    const volumeRatio = recentAvgVolume / olderAvgVolume

    if (volumeRatio >= 1.5) {
      trendingCandidates.push({
        stock,
        volumeRatio,
      })
    }
  }

  // 取引高増加率が高い順にソート
  trendingCandidates.sort((a, b) => b.volumeRatio - a.volumeRatio)

  // Top 3を選出
  const topTrending = trendingCandidates.slice(0, 3)

  const results: FeaturedStock[] = topTrending.map((candidate, idx) => {
    const marketCapLabel = candidate.stock.marketCap >= 10000
      ? `${(candidate.stock.marketCap / 10000).toFixed(1)}兆円`
      : `${candidate.stock.marketCap.toFixed(0)}億円`
    const reason = `直近3日で取引が活発になっている注目銘柄です（取引高${candidate.volumeRatio.toFixed(1)}倍、時価総額${marketCapLabel}）`

    return {
      stockId: candidate.stock.id,
      category: "trending",
      categoryPosition: idx + 1,
      reason,
      score: Math.min(100, Math.floor(candidate.stock.marketCap / 100)), // 時価総額ベースのスコア
    }
  })

  console.log(`Trending: ${results.length} stocks selected`)
  return results
}

async function saveDailyFeaturedStocks(featuredStocks: FeaturedStock[]): Promise<void> {
  if (featuredStocks.length === 0) {
    console.log("No stocks to save")
    return
  }

  const today = dayjs.utc().startOf("day").toDate()

  // 既存データを削除（今日の日付）
  const deleted = await prisma.dailyFeaturedStock.deleteMany({
    where: { date: today },
  })
  console.log(`Deleted ${deleted.count} existing records for ${today.toISOString().split("T")[0]}`)

  // 新しいデータを挿入（positionは全体通し番号1-9）
  for (let idx = 0; idx < featuredStocks.length; idx++) {
    const fs = featuredStocks[idx]
    await prisma.dailyFeaturedStock.create({
      data: {
        date: today,
        stockId: fs.stockId,
        category: fs.category,
        position: idx + 1, // 全体通し番号
        reason: fs.reason,
        score: fs.score,
      },
    })
  }

  console.log(`Saved ${featuredStocks.length} featured stocks for ${today.toISOString().split("T")[0]}`)
}

async function main(): Promise<void> {
  console.log("=".repeat(60))
  console.log("DailyFeaturedStock Generation")
  console.log("=".repeat(60))
  console.log(`Time: ${new Date().toISOString()}`)

  try {
    // 銘柄と株価データを取得
    console.log("\nFetching stocks and price data...")
    const stocks = await getStocksWithPrices()

    if (stocks.length === 0) {
      console.log("No stocks with sufficient price data. Exiting.")
      return
    }

    // 各カテゴリの銘柄を抽出
    console.log("\nCalculating featured stocks...")

    const surgeStocks = calculateSurgeStocks(stocks)
    const stableStocks = calculateStableStocks(stocks)
    const trendingStocks = calculateTrendingStocks(stocks)

    // 結果を結合
    const allFeatured = [...surgeStocks, ...stableStocks, ...trendingStocks]

    if (allFeatured.length === 0) {
      console.log("No stocks matched criteria today")
      return
    }

    // データベースに保存
    console.log("\nSaving to database...")
    await saveDailyFeaturedStocks(allFeatured)

    // サマリー表示
    console.log("\n" + "=".repeat(60))
    console.log("DailyFeaturedStock generation completed")
    console.log("=".repeat(60))
    console.log(`Total featured stocks: ${allFeatured.length}`)
    console.log(`  - Surge: ${surgeStocks.length}`)
    console.log(`  - Stable: ${stableStocks.length}`)
    console.log(`  - Trending: ${trendingStocks.length}`)
    console.log("=".repeat(60))
  } catch (error) {
    console.error(`Error: ${error}`)
    process.exit(1)
  } finally {
    await prisma.$disconnect()
  }
}

main()

export {}
